protocol/contracts/oud/OudRedeemer.sol

```diff
@@ -0,0 +1,145 @@
+    return tpi;
+  }
+
+  // @dev Allows the tpi to be adjusted.
+  // @param 'value' is the new TPI
+  function setTreasuryPriceIndex(uint256 value) external override onlyOwner {
+    tpi = value;
+  }
+
+  function setStableCoin(address newStable) external override onlyOwner {
+    if (newStable == address(0)) revert AddressZero();
+    stable = IERC20(newStable);
+  }
+
+  // @dev Decimal precision for 'tpi', 9880 == $0.988, precision = 4
+  function TPI_PRECISION() external pure returns (uint256) {
+    return tpiDecimals;
+  }
+
+  /**
+   * @notice Provides a quote for 'oudAmount' with relevant token addresses:
+   *  'oudAmount' provided, 'stable' token required, 'templeAmount' that will be minted to caller
+   * @param oudAmount amount of Oud that will be redeemed
+   * @return _oudAmount amount of Oud that wil be redeemed
+   * @return _oudTokenAddress address of the Oud token
+   * @return _stableAmount amount of Stable token required at tpi
+   * @return _stableAddress address of the Stable token in use
+   * @return _templeAmount amount of Temple that will be minted
+   * @return _templeTokenAddress address of the Temple token
+   */
+  function getQuote(
+    uint256 oudAmount
+  )
+    external
+    view
+    returns (
+      uint256 _oudAmount,
+      address _oudTokenAddress,
+      uint256 _stableAmount,
+      address _stableAddress,
+      uint256 _templeAmount,
+      address _templeTokenAddress
+    )
+  {
+    return (
+      oudAmount,
+      oudToken,
+      _getStableAmount(oudAmount),
+      address(stable),
+      oudAmount,
+      templeToken
+    );
+  }
+
+  // @notice Redeems Oud and Stable token (@ tpi) for Temple
+  // @dev requires allowances for Oud and Stable Token
+  // @param 'oudAmount' The amount of Oud to be redeemed for Temple
+  // @returns 'templeAmount' The amount of Temple that has been minted to sender.
+  function redeem(uint256 oudAmount) external returns (uint256 templeAmount) {
+    if (oudAmount == 0) revert RedeemAmountZero();
+    address account = msg.sender;
+    uint256 _stableAmount;
+    _stableAmount = _getStableAmount(oudAmount);
+    IOudToken(oudToken).burn(account, oudAmount);
+    SafeERC20.safeTransferFrom(stable, account, address(this), _stableAmount);
+    SafeERC20.safeIncreaseAllowance(stable, depositStableTo, _stableAmount);
+    Oud__ITempleLineOfCredit(depositStableTo).depositReserve(
+      stable,
+      address(this),
+      _stableAmount
+    );
+
+    // @dev Conversion is 1 OUD + (Stable*TPI) = 1 Temple thus Temple Amount == Oud Amount.
+    Oud__ITempleERC20Token(templeToken).mint(account, oudAmount);
+    emit OudRedeemed(account, oudAmount, tpi, oudAmount);
+    return oudAmount;
+  }
+
+  function _getStableAmount(
+    uint256 oudAmount
+  ) internal view returns (uint256 stableAmount) {
+    return (oudAmount * tpi) / (10 ** tpiDecimals);
+  }
+
+  event OudRedeemed(
+    address account,
+    uint256 OudRedeemed,
```

Comment:
should be lower case.

And perhaps should also have the stable amount here too?

Reply:
[2eba8d9](https://github.com/TempleDAO/temple/pull/760/commits/2eba8d9c319da00aa83f35f11e2f2d2123b34865)